### PR TITLE
Add `dot` method for symmetric matrices

### LINKED ISSUE #647: Faster dot product for symmetric matrices
I want to get an inner product between two symmetric matrices, and `dot` seems to dispatch to a pretty slow method.
E.g.

```julia
using BenchmarkTools, LinearAlgebra
n = 5000;
A = Symmetric(randn(n, n));
B = Symmetric(randn(n, n));
function symdot(A::Symmetric{T, Matrix{T}}, B::Symmetric{T, Matrix{T}}) where {T}
    ret = zero(T)
    m = size(A, 2)
    @inbounds for j in 1:m
        for i in 1:(j - 1)
            ret += A[i, j] * B[i, j] * 2
        end
        ret += A[j, j] * B[j, j]
    end
    return ret
end

julia> @benchmark dot($A, $B)
benchmark symdot($A, $B)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     371.083 ms (0.00% GC)
  median time:      371.527 ms (0.00% GC)
  mean time:        371.907 ms (0.00% GC)
  maximum time:     377.313 ms (0.00% GC)
  --------------
  samples:          14
  evals/sample:     1

julia> @benchmark symdot($A, $B)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     39.275 ms (0.00% GC)
  median time:      39.359 ms (0.00% GC)
  mean time:        39.378 ms (0.00% GC)
  maximum time:     39.814 ms (0.00% GC)
  --------------
  samples:          127
  evals/sample:     1
```

Any chance of getting a faster dot product in?

This is very specific, but related to https://github.com/JuliaLang/LinearAlgebra.jl/issues/496 and https://github.com/JuliaLang/LinearAlgebra.jl/issues/334.

### LINKED ISSUE #32730: Faster dot product for symmetric matrices
I want to get an inner product between two symmetric matrices, and `dot` seems to dispatch to a pretty slow method.
E.g.

```julia
using BenchmarkTools, LinearAlgebra
n = 5000;
A = Symmetric(randn(n, n));
B = Symmetric(randn(n, n));
function symdot(A::Symmetric{T, Matrix{T}}, B::Symmetric{T, Matrix{T}}) where {T}
    ret = zero(T)
    m = size(A, 2)
    @inbounds for j in 1:m
        for i in 1:(j - 1)
            ret += A[i, j] * B[i, j] * 2
        end
        ret += A[j, j] * B[j, j]
    end
    return ret
end

julia> @benchmark dot($A, $B)
benchmark symdot($A, $B)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     371.083 ms (0.00% GC)
  median time:      371.527 ms (0.00% GC)
  mean time:        371.907 ms (0.00% GC)
  maximum time:     377.313 ms (0.00% GC)
  --------------
  samples:          14
  evals/sample:     1

julia> @benchmark symdot($A, $B)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     39.275 ms (0.00% GC)
  median time:      39.359 ms (0.00% GC)
  mean time:        39.378 ms (0.00% GC)
  maximum time:     39.814 ms (0.00% GC)
  --------------
  samples:          127
  evals/sample:     1
```

Any chance of getting a faster dot product in?

This is very specific, but related to https://github.com/JuliaLang/julia/issues/25565 and https://github.com/JuliaLang/julia/issues/16573.

### PR DESCRIPTION
**Pull request by goggle**
This PR adds a `dot` method for symmetric matrices, which makes the calculation of the dot product of two symmetric matrices significantly faster (see benchmarks below). It intends to fix JuliaLang/LinearAlgebra.jl#647.

To perform the benchmarks, this code was used:
```julia
using LinearAlgebra
using BenchmarkTools

n = 10_000
X = randn(n, n)
Y = randn(n, n)

A = Symmetric(X)  # or A = Symmetric(X, :L)
B = Symmetric(Y)  # or B = Symmetric(Y, :L)
@btime dot(A, B)
```

These are the results:

| A.uplo | B.uplo | Julia 1.1    | PR            |
| :--------: | :--------: | -------------: | -------------: |
| 'U'       | 'U'       | 1.4 s         | 72 ms        |
| 'U'       | 'L'       | 1.33 s        | 537 ms      |
| 'L'       | 'U'       | 1.33 s        | 538 ms      |
| 'L'        | 'L'        | 1.4 s         | 72 ms        |

### CODE REVIEW

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl]: Actually, I wonder if it should be
```julia
dotprod += 2real(dot(A.data[i, j], B.data[i, j]))
```
because, for a symmetric block matrix `X` with complex matrix blocks, you would want that `dot(X, X)` returns a real number.

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetric.jl]: Both `A` and `B` are square by type, so it would suffice to get `size(_, 1)` and do only one comparison below... and use only `m` and not both `m` and `n` in the loops.

**Review by dkarrasch — CHANGES_REQUESTED** [review]
This looks good, but I guess a few changes are required for block matrices, taking into account that `dot` acts recursively.

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (goggle) [reply stdlib/LinearAlgebra/src/symmetric.jl]: Yes, I've noticed this too by looking up other implementation of currently available `dot` methods: They all use a *inner* `dot` method like you suggested. I'm not familiar with block matrices in the Julia standard library yet, but I will look this up.

I guess it will also make sense to add a test for symmetric block matrices.

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (goggle) [reply stdlib/LinearAlgebra/src/symmetric.jl]: By 

> relax the "same type" requirement

you mean that this method should look like this?
```julia
function dot(A::Symmetric{T, Matrix{T}}, B::Symmetric{S, Matrix{S}}) where {T, S}
```

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl]: I think the main guideline to follow is to make/keep `dot` recursive, something I also realized only while developing #32739. Whenever you're tempted to write out a product and conjugate one of the factors, just use `dot` instead and you'll get it right. At the very bottom of the hierarchy when it reaches numbers, this is what it's going to do anyway.

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetric.jl]: I would simply say
```julia
function dot(A::Symmetric, B::Symmetric)
```
Whatever the eltype is, you later request that they have (on the highest level) the same size, and then you go through the elements and `dot` them. If there is a mismatch in hierarchy, the subsequent `dot`s are going to realize that, and throw an error. `Matrix` is in many cases rather restrictive, so you couldn't pass a view, or a reshape.

Comment (goggle) [reply stdlib/LinearAlgebra/src/symmetric.jl]: Ah ok, thanks. I will try this.

**Comment by goggle** [comment]
I have done the requested changes by @dkarrasch.

I don't quite understand the possible problem with block matrices. I didn't find them in the Julia stdlib, so I guess you mean the package [BlockArrays.jl](https://github.com/JuliaArrays/BlockArrays.jl) @dkarrasch? If you think there is an issue, feel free to elaborate.

**Comment by dkarrasch** [comment]
No, I meant matrices of matrices. Like
```julia
X = [rand(ComplexF64, 10, 10) for _ in 1:3, _ in 1:3]
dot(Symmetric(X), Symmetric(X)) # should give a real, non-negative number
```

**Comment by antoine-levitt** [comment]
Could this also support Hermitian?

I think it'd be good to add more comprehensive tests, including the four code paths (L/R), complex and block matrices: mistakes in functions like these might not be trivial to spot, and could have very bad consequences.

**Comment by goggle** [comment]
@dkarrasch Thanks for the clarification, I'm going to check this!

@antoine-levitt Sure, I can add a similar method for `Hermitian`, but I would prefer to do this in a followup-PR. I agree on improving the tests.

**Comment by chriscoey** [comment]
@goggle thanks for this! It speeds up my code. 
I agree with@antoine-levitt's points. Hermitian should be an easy generalization.

**Comment by goggle** [comment]
There is indeed an issue with block matrices.
Let's consider consider this example:
```julia
using LinearAlgebra
a = [1 0 ; 1 0]
b = [-1 1; 0 1]
c = [-1 -1 ; 0 0]
d = [-1 2 ; -1 0]
A = [[a] [c] ; [b] [d]]
symA = Symmetric(A)
```
`symA` is now a matrix consisting of four (2x2)-blocks, such that `symA` is symmetric when considered as a 4x4 matrix. In the `symA.data` fields we have the origin data (the arrays `a`, `b`, `c` and `d`), so my method produces a wrong answer: `dot(symA, symA) = 14` on Julia 1.1.1 vs. `dot(symA, symA) = 12` on  https://github.com/JuliaLang/julia/pull/32827/commits/738b89aedd6a3a9a76d1537e46dd7d162e7039c2.

I will replace all the `A.data` and `B.data` by `A` and `B`. This should produce correct results, but might be slightly slower, so I will also rerun the benchmark.

**Comment by goggle** [comment]
Alright, so I avoid to access the data directly, which should solve the issues with symmetric block matrices. As expected, it makes it slightly slower than the previous version. Here are the updated benchmarks:

| A.uplo | B.uplo | Julia 1.1    | https://github.com/JuliaLang/julia/pull/32827/commits/bae07539f95122e1d3019c4bfb746b04feae5804            | https://github.com/JuliaLang/julia/pull/32827/commits/dcc1aaee7eb2b42b4df8397c31bb87b6c412fb0b |
| :--------: | :--------: | -------------: | -------------: |--------------:
| 'U'       | 'U'       | 1.4 s         | 72 ms        | 109 ms |
| 'U'       | 'L'       | 1.33 s        | 537 ms      | 602 ms |
| 'L'       | 'U'       | 1.33 s        | 538 ms      | 608 ms |
| 'L'        | 'L'        | 1.4 s         | 72 ms        | 100 ms |

I will extend the tests tomorrow.

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (chriscoey) [inline stdlib/LinearAlgebra/src/symmetric.jl]: Missing a space between indices

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetric.jl]: To handle nested arrays, this needs to be 
```julia
dotprod += dot(symmetric(A.data[j, j], sym_uplo(A.uplo), symmetric(B.data[j, j], sym_uplo(B.uplo))
```
or maybe simply
```julia
dotprod += dot(A[j, j], B[j, j])
```
this will then take care of the symmetrization of diagonal elements.

And similarly below...

File: stdlib/LinearAlgebra/test/symmetric.jl
Comment (dkarrasch) [inline stdlib/LinearAlgebra/test/symmetric.jl]: For the nested matrices, you could simply have a couple of tests like
```julia
X = [rand(eltyXYZ, 10, 10) for _ in 1:3, _ in 1:3]
@test dot(Symmetric(X), Symmetric(X)) ≈ dot(Matrix(Symmetric(X)), Matrix(Symmetric(X)))
```
There is no need to create block symmetric matrices by hand, simply let the wrapper do, and afterwards get rid of the wrapper by `Matrix`.

**Review by dkarrasch — CHANGES_REQUESTED** [review]
I don’t see any need to restrict to numbers. The remaining issues for nested arrays are easily fixed. What was missing is using the implicit symmetry of diagonal elements, which may not be already contained in the `data` field. For the `Hermitian` version of this (not sure if we would also add mixed versions), you will need to replace the off-diagonal contributions by
```julia
dotprod += 2real(dot(A.data[i, j], B.data[i, j]))
```
otherwise this should be the same.

**Comment by goggle** [comment]
Thanks for these detailed comments @dkarrasch, this helps a lot. I would also prefer to have a generic implementation `dot(A::Symmetric, B::Symmetric)` for the dot product of symmetric matrices, but I have run into some issues that still persist.

Here is the modified dot method with your suggestions from https://github.com/JuliaLang/julia/pull/32827#discussion_r313996019 and https://github.com/JuliaLang/julia/pull/32827#discussion_r314045142 applied:
```julia
using LinearAlgebra

function mdot(A::Symmetric, B::Symmetric)
    n = size(A, 2)
    if n != size(B, 2)
        throw(DimensionMismatch("A has dimensions $(size(A)) but B has dimensions $(size(B))"))
    end

    dotprod = zero(dot(first(A), first(B)))
    @inbounds if A.uplo == 'U' && B.uplo == 'U'
        for j in 1:n
            for i in 1:(j - 1)
                dotprod += 2 * dot(A.data[i, j], B.data[i, j])
            end
            dotprod += dot(A[j, j], B[j, j])
        end
    elseif A.uplo == 'L' && B.uplo == 'L'
        for j in 1:n
            dotprod += dot(A[j, j], B[j, j])
            for i in (j + 1):n
                dotprod += 2 * dot(A.data[i, j], B.data[i, j])
            end
        end
    elseif A.uplo == 'U' && B.uplo == 'L'
        for j in 1:n
            for i in 1:(j - 1)
                dotprod += 2 * dot(A.data[i, j], B.data[j, i])
            end
            dotprod += dot(A[j, j], B[j, j])
        end
    elseif A.uplo == 'L' && B.uplo == 'U'
        for j in 1:n
            dotprod += dot(A[j, j], B[j, j])
            for i in (j + 1):n
                dotprod += 2 * dot(A.data[i, j], B.data[j, i])
            end
        end
    end
    return dotprod
end
```
I call it `mdot` here. Let's take the block matrix from https://github.com/JuliaLang/julia/pull/32827#issuecomment-520195202:
```
a = [1 0 ; 1 0]
b = [-1 1; 0 1]
c = [-1 -1 ; 0 0]
d = [-1 2 ; -1 0]
A = [[a] [c] ; [b] [d]]
sau = Symmetric(A, :U)
sal = Symmetric(A, :L)
```
Now I get a wrong result with `mdot`:
```
julia> mdot(sau, sal)
-2
```
It should be:
```
julia> dot(sau, sal)
0
```
That's exactly the same issue that I could not resolve earlier...
Do you have any solution for this?

**Comment by dkarrasch** [comment]
Your example seems to be fixed by
```julia
using LinearAlgebra

function mdot(A::Symmetric, B::Symmetric)
    n = size(A, 2)
    if n != size(B, 2)
        throw(DimensionMismatch("A has dimensions $(size(A)) but B has dimensions $(size(B))"))
    end

    dotprod = zero(dot(first(A), first(B)))
    @inbounds if A.uplo == 'U' && B.uplo == 'U'
        for j in 1:n
            for i in 1:(j - 1)
                dotprod += 2 * dot(A.data[i, j], B.data[i, j])
            end
            dotprod += dot(A[j, j], B[j, j])
        end
    elseif A.uplo == 'L' && B.uplo == 'L'
        for j in 1:n
            dotprod += dot(A[j, j], B[j, j])
            for i in (j + 1):n
                dotprod += 2 * dot(A.data[i, j], B.data[i, j])
            end
        end
    elseif A.uplo == 'U' && B.uplo == 'L'
        for j in 1:n
            for i in 1:(j - 1)
                dotprod += 2 * dot(A.data[i, j], transpose(B.data[j, i]))
            end
            dotprod += dot(A[j, j], B[j, j])
        end
    elseif A.uplo == 'L' && B.uplo == 'U'
        for j in 1:n
            dotprod += dot(A[j, j], B[j, j])
            for i in (j + 1):n
                dotprod += 2 * dot(transpose(A.data[i, j]), B.data[j, i])
            end
        end
    end
    return dotprod
end
```
Note the two `transpose`s in the mixed cases. They are consistent with the `getindex` method for `Symmetric`s.

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetric.jl]: It should be `transpose(B.data[j, j])` here and ...

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetric.jl]: ... some `transpose` here, but I’m not sure which (or if it makes a different at all whether at `A` or `B`).

**Review by dkarrasch — COMMENTED** [review]
There are `transpose`s missing in the mixed cases, please double-check with the `getindex` code.

**Comment by goggle** [comment]
Ok, I think I could solve all the remaining issues, thanks @dkarrasch.
Is there anything else?

**Comment by dkarrasch** [comment]
LGTM. How about adding the Hermitian-Hermitian case also? You would need to replace all `transpose`s by `adjoint`s, and maybe add the `real` in the `2dot(...)` lines. You could put all your tests in a loop over `transform in (Symmetric, Hermitian)`, and replace all instances of `Symmetric` by `transform`, so adopting the tests should be easy.

**Comment by goggle** [comment]
I have added a second method for the Hermitian-Hermitian case and have added tests for the Hermitian case as well.

File: stdlib/LinearAlgebra/src/symmetric.jl
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetric.jl]: This is the last remaining combination, so this could be shortened to simply `else` without condition. Similarly in the Hermitian case.

**Review by dkarrasch — APPROVED** [review]
Just a trivial comment, otherwise looks good to me. Could you maybe post an update (or a reminder) of the performance improvements by this PR?

**Comment by chriscoey** [comment]
Thanks for being so thorough on this

**Comment by goggle** [comment]
Here are some final benchmarks.
```julia
using BenchmarkTools
using LinearAlgebra
n = 5000;

Au = Symmetric(randn(n,n));
Al = Symmetric(randn(n,n), :L);
Bu = Symmetric(randn(n,n));
Bl = Symmetric(randn(n,n), :L);
Hu = Hermitian(randn(n,n) + randn(n,n)*im);
Hl = Hermitian(randn(n,n) + randn(n,n)*im, :L);
Il = Hermitian(randn(n,n) + randn(n,n)*im, :L);
Iu = Hermitian(randn(n,n) + randn(n,n)*im);
```

| Command                 | Julia 1.1 | Commit https://github.com/JuliaLang/julia/pull/32827/commits/8d75d17644674572dc4a0e8ff3580c2103653cd9 |
|:---------------------------|------------:|------------------:|
| `dot(Au, Bu)`             | 318 ms    | 18 ms             |
| `dot(Au, Bl)`               | 278 ms   |  114 ms          |
| `dot(Al, Bu)`               | 308 ms   | 112 ms           |
| `dot(Al, Bl)`                 | 388 ms   | 18 ms            |
| `dot(Hu, Iu)`             | 388 ms     | 34 ms            |
| `dot(Hu, Il)`              | 418 ms     | 135 ms          |
| `dot(Hl, Iu)`              | 383 ms     | 135 ms          |
| `dot(Hl, Il)`                | 385 ms     | 35 ms           |

@kshyatt Maybe you could add the `performance` label to this PR. Thanks!

**Comment by goggle** [comment]
Any news on this one? Does it need a second review? How can I proceed?

**Comment by dkarrasch** [comment]
Like a good piece of meat, this needs to hang a little... 😉 until someone with merge rights finds some time to review, and merges.

**Comment by dkarrasch** [comment]
@goggle Could you please rebase onto master/resolve the merge conflict, and then I'll merge!

**Comment by goggle** [comment]
I have rebased the branch to resolve the conflicts.
@dkarrasch Let me know if I should also squash the commits.

**Comment by dkarrasch** [comment]
Don't worry about squashing, unless this is more convenient for you while developing. Squashing will be done upon merging at the end anyway. I'll wait for tests to pass...

**Comment by dkarrasch** [comment]
Seems like one `end` got lost in the merge process. Add one after line 378, and then I expect it to work just fine.

**Comment by lkapelevich** [comment]
Thanks @goggle!

**Comment by goggle** [comment]
Also thanks to @dkarrasch for the great support!

**Pull request merged**
